File: Audio effects in Matlab/3.3.2 chorus/chorus.m
clear all;
close all;
clc;

[x,Fs] = wavread('..\guitar.wav');
N = length(x);
n=0:length(x)-1;
time=n/Fs;


dry = 1;
wet = 0.5;
delay = round(0.02*Fs)    %20 ms       
amount = round(0.01*Fs)   %10 ms
advance=1.43;
w0=2*pi*(advance/Fs);
n=1:1:N;
sinus = round(delay + amount*sin(w0*n));
y=zeros(size(x));
for n = delay + amount:1:N - delay - amount
    y(n) = dry*x(n) + wet*x(n-sinus(n));
end

fid=fopen('x.dat','w','b');
fwrite(fid,x.*2^15,'int16');
fclose(fid);
fid=fopen('sinus.dat','w','b');
fwrite(fid,sinus,'int16');
fclose(fid);

display('APASA O TASTA');
pause


fid=fopen('y.dat','r','b');
y_cw=fread(fid,N,'int16');
fclose(fid);
y_cw=y_cw'/(2^15);

% reprezentarea lui x in timp (in secunde)
figure(1),plot(time,x),grid,title('Semnalul de test'),xlabel('Timp (s)');
%sound(x, Fs);
figure(2),plot(time,y),grid,title('Semnalul chorus'),xlabel('Timp (s)');
sound(y, Fs);
figure(3),plot(y_cw),grid,title('Semnalul chorus'),xlabel('Timp (s)');
sound(y_cw, Fs);

File: Audio effects in Matlab/3.3.2 chorus/Chorus.c
#include <prototype.h> 
#include <stdio.h> 
#define	DataBlockSize	110		/* size of data block to process */
#define BlockLength    1000		/* number of data blocks to process */
#define	N	662					/* delay +amount*/

int main()
{
Word16 x[DataBlockSize],y[DataBlockSize],sinus[DataBlockSize],delay[N];
short n,k;
long unsigned index=0;
Word16 wet = WORD16(0.5);
//Word16 dry = WORD16(1);

FILE *fpx,*fpy,*fpsinus;

fpx=fopen("x.dat","r+b");
if (!fpx)
    printf("\nNu s-a deschis");

fpsinus=fopen("sinus.dat","r+b");
if (!fpsinus)
    printf("\nNu s-a deschis");

fpy=fopen("y.dat","w+b");
if (!fpy)
    printf("\nNu s-a deschis");




for (k=0; k<BlockLength; k++)
{
	fread(x,sizeof(Word16),DataBlockSize,fpx);
	fread(sinus,sizeof(Word16),DataBlockSize,fpsinus);
	
	for (n=0; n<DataBlockSize; n++)
	{
		y[n]=mac_r(L_deposit_h(x[n]),delay[(index+N-sinus[n])%N],wet);// delay[(index+N-d[n])%N]=x(n-d(n))
		delay[index%N]=x[n];

		index++;
	}

	fwrite(y,sizeof(Word16),DataBlockSize,fpy);
}

fclose(fpx);
fclose(fpy);

return(0);
}
